test/shared/src/main/scala/zio/test/TestAnnotations.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import zio.{ Ref, UIO, ZIO }
+
+/**
+ * The `TestAnnotations` trait provides access to an annotation map that tests
+ * can add arbitrary annotations to. Each annotation consists of a string
+ * identifier, an initial value, and a function for combining two values.
+ * Annotations form monoids and you can think of `TestAnnotations` as a more
+ * structured logging service or as a super polymorphic version of the writer
+ * monad effect.
+ */
+trait TestAnnotations {
+  val testAnnotations: TestAnnotations.Service[Any]
+}
+
+object TestAnnotations {
+
+  trait Service[R] {
+    def annotate[V](key: TestAnnotation[V], value: V): ZIO[R, Nothing, Unit]
+    def get[V](key: TestAnnotation[V]): ZIO[R, Nothing, V]
+    val testAnnotationMap: ZIO[R, Nothing, TestAnnotationMap]
```

Comment:
Do we have to expose this `map`? It would be nice to make this as tiny a surface area as possible, which means `annotate` and `get`.

Reply:
If we want executing a spec to give you the annotation map we need a way to retrieve the annotation map. We could make it package private, though that would mean any user implementing their own test executor would not be able to access the annotation map.